presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,191 @@
+
+@ThreadSafe
+public class ClusterMemoryLeakDetector
+{
+    private static final Logger log = Logger.get(ClusterMemoryLeakDetector.class);
+
+    // It may take some time to remove a query's memory reservations from the worker nodes, that's why
+    // we check to see whether some time has passed after the query finishes to claim that it is leaked.
+    private static final int DEFAULT_LEAK_CLAIM_DELTA_SEC = 60;
+
+    // By default, leakedQueries will be a WeakHashMap where the keys are leaked query IDs and the values are TRUE.
+    // This is for just being safe in case the map gets too large.
+    @GuardedBy("this")
+    private final LeakedQueries leakedQueries;
+
+    // We don't rely on the size of the weak leakedQueries map for the number of leaked queries as we want an exact counter.
+    private final AtomicInteger numberOfLeakedQueries = new AtomicInteger();
+
+    static ClusterMemoryLeakDetector createMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new DefaultLeakedQueries());
+    }
+
+    @VisibleForTesting
+    static ClusterMemoryLeakDetector createTestMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new TestLeakedQueries());
+    }
+
+    private ClusterMemoryLeakDetector(LeakedQueries leakedQueries)
+    {
+        this.leakedQueries = requireNonNull(leakedQueries, "leakedQueries is null");
+    }
+
+    /**
+     * @param allQueries All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(List<QueryInfo> allQueries, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(allQueries);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> shouldClaimQueryAsLeaked(allQueries, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean shouldClaimQueryAsLeaked(List<QueryInfo> allQueries, QueryId queryId)
+    {
+        boolean queryExists = false;
+        QueryState queryState = null;
+        DateTime queryEndTime = null;
+
+        for (QueryInfo queryInfo : allQueries) {
```

Comment:
Is this O(n^2) with regards to number of queries?

Reply:
> This particular loop should be O(n), no?

Yes, but method seems being called O(n) times.

This won't be problem unless you have multiple concurrent queries. However, why not `com.google.common.collect.Maps#uniqueIndex` the `allQueries` (or perhaps Stream-based equivalent) before calling here?